homeassistant/components/knx/__init__.py

```diff
@@ -235,23 +256,56 @@ async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
 
     async def reload_service_handler(service_call: ServiceCall) -> None:
         """Remove all KNX components and load new ones from config."""
+        await hass.config_entries.async_reload(entry.entry_id)
 
-        # First check for config file. If for some reason it is no longer there
-        # or knx is no longer mentioned, stop the reload.
-        config = await async_integration_yaml_config(hass, DOMAIN)
-        if not config or DOMAIN not in config:
-            return
+    async_register_admin_service(
+        hass, DOMAIN, SERVICE_RELOAD, reload_service_handler, schema=vol.Schema({})
+    )
 
-        await asyncio.gather(
-            *(platform.async_reset() for platform in async_get_platforms(hass, DOMAIN))
-        )
-        await knx_module.xknx.stop()
+    return True
 
-        await async_setup(hass, config)
 
-    async_register_admin_service(
-        hass, DOMAIN, SERVICE_RELOAD, reload_service_handler, schema=vol.Schema({})
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unloading the KNX platforms."""
+    knx_module: KNXModule = hass.data[DOMAIN]
+    for exposure in knx_module.exposures:
+        exposure.shutdown()
+
+    unload_ok = await hass.config_entries.async_unload_platforms(
+        entry,
+        [
+            platform.value
+            for platform in SupportedPlatforms
+            if platform.value in hass.data[DATA_KNX_CONFIG]
+        ],
     )
+    if unload_ok:
+        await knx_module.stop()
+        hass.data[DOMAIN] = None
+
+    return unload_ok
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Start the KNX integration."""
+    conf: ConfigType | None = config.get(DOMAIN)
+
+    if conf is None:
+        # If we have a config entry, setup is done by that config entry.
+        # If there is no config entry, this should fail.
+        return bool(hass.config_entries.async_entries(DOMAIN))
+
+    conf = dict(conf)
+
+    hass.data[DATA_KNX_CONFIG] = conf
+
+    # Only import if we haven't before.
```

Comment:
You don't want to make this update the config entry ?

Reply:
I thought we can do a one time migration and then deprecate the connection yaml in a follow up PR.

Reply:
I talked to farmio, we think it's best if we leave it like that and then later (half a year - year) remove the YAML configs for the connection handling altogether.